### PR TITLE
Provide IPv6 support

### PR DESCRIPTION
**Pull request by pmichali**
Adding a WIP change (on top of the fix-1.8+ branch) that will allow DinD to be used on an IPv6 only Kubernetes cluster.

This uses IPv6 addresses for all pods, and makes use of DNS64 and NAT64 containers, to convert external IPs to IPv4 embedded IPv6 addresses.

An IP_MODE environment variable is added to select between IPv4 (default) and IPv6. Additional environment variables can be set to customize DinD subnet, service subnet CIDR, and prefix to use for DNS64.

To use, must use the customer Kubernetes repo in the README.md, as it has IPv6 patches that are in-flight (or fixing bugs).

Notes:
- Uses /80 for pod subnet size.
- Requires building local copy, due to changes to DinD code.
- All pod subnets are "inside" of the DinD /64 subnet (same as what IPv4 does).
- Includes fix to restore controller manifest, when failure to shutdown controller/proxy pods during snapshot.
- Relies on both a custom Kubernetes repo (until commits are upstreamed), and kube-dns image (referenced in manifest of custom Kubernetes repo).
- Could be customized to use NAT66 container as an option to allow access to external IPv6 IPs.
- Future could add dual-stack support.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -123,6 +123,48 @@ The first `dind/dind-cluster.sh up` invocation can be slow because it
...
+$ export DIND_IMAGE=mirantis/kubeadm-dind-cluster:local
+```
+
+Next, clone this custom Kubernetes repo, which has a set of patches on
+top of a recent master release:
+
+```shell
+$ git clone https://github.com/pmichali/kubernetes.git
+$ cd kubernetes
+$ git checkout -b kubernetes_ipv6
```
Comment (danehans) [inline README.md]: Instead of using your fork, I suggest using upstream and listing the necessary PRs. As the PRs merge, doc update PRs can be submitted.

File: README.md [hunk 2]
```diff
@@ -123,6 +123,48 @@ The first `dind/dind-cluster.sh up` invocation can be slow because it
...
+## IPv6 Mode (experimental)
+To run Kubernetes in IPv6 only mode, the following steps can be performed
+in a Kubernetes source setup. First, build a new kubeadm-dind-cluster
+image, from the DinD area:
+
+```shell
+$ cd ~/dind
+
+$ build/build-local.sh
+$ export DIND_IMAGE=mirantis/kubeadm-dind-cluster:local
```
Comment (danehans) [inline README.md]: Is this build step needed if this PR gets merged? If so, I suggest removing it.

File: config.sh [hunk 3]
```diff
@@ -1,6 +1,3 @@
-# DIND subnet (/16 is always used)
-DIND_SUBNET=10.192.0.0
-
```
Comment (danehans) [inline config.sh]: Why does din subnet need to be removed from the config?

File: dind-cluster.sh [hunk 4]
```diff
@@ -368,7 +402,12 @@ function dind::ensure-binaries {
 
 function dind::ensure-network {
   if ! docker network inspect kubeadm-dind-net >&/dev/null; then
-    docker network create --subnet="${DIND_SUBNET}/16" kubeadm-dind-net >/dev/null
+    local v6settings=""
+    if [[ ${IP_MODE} = "ipv6" ]]; then
+      # Need second network
```
Comment (danehans) [inline dind-cluster.sh]: Can you add to the comment why a 2nd network is needed, i.e. for nat64.

File: README.md (hunk 1 above)
Comment (pmichali) [reply README.md]: I wanted to have an easy way for people to test, rather than forcing them to patch in a bunch of PRs. I do list the PRs in my blog (http://blog.michali.net/2017/08/09/ipv6-support-for-docker-in-docker/).

Would like to know how we want to manage this. Also, I have one change for a temp hack for an issue where there is no PR yet.

File: dind-cluster.sh [hunk 5]
```diff
@@ -557,12 +666,27 @@ function dind::init {
...
+    local pod_net_cidr=""
+    # TODO: May want to specify each of the plugins that require --pod-network-cidr
+    if [[ ${CNI_PLUGIN} != "bridge" ]]; then
+      pod_net_cidr="podSubnet: \"${POD_NETWORK_CIDR}\""$'\n'"  "
+    fi
+    local bind_address="0.0.0.0"
+    if [[ $IP_MODE = "ipv6" ]]; then
+	bind_address="::"
+    fi
     docker exec -i kube-master /bin/sh -c "cat >/etc/kubeadm.conf" <<EOF
```
Comment (pmichali) [inline dind-cluster.sh]: I'm wondering if we should built a temp file with kubeadm.conf, and then docker cp it to the master, rather than trying to build on the fly.

File: config.sh (hunk 3 above)
Comment (pmichali) [reply config.sh]: Because DinD will force this as the subnet, and we cannot override, unless we manually set the environment variable EMBEDDED_CONFIG. Instead, I remove it from here, as set it in dind-cluster.sh, allowing the user to override, if desired, and to set a default that is IP mode specific.

File: dind-cluster.sh (hunk 4 above)
Comment (pmichali) [reply dind-cluster.sh]: Sure will do.

File: dind-cluster.sh [hunk 6]
```diff
@@ -46,10 +46,44 @@ if [[ ! ${EMBEDDED_CONFIG:-} ]]; then
...
+else
+    DIND_SUBNET="${DIND_SUBNET:-10.192.0.0}"
+    dind_ip_base="$(echo "${DIND_SUBNET}" | sed 's/0$//')"
+    KUBE_RSYNC_ADDR="${KUBE_RSYNC_ADDR:-127.0.0.1}"
+    SERVICE_CIDR="${SERVICE_CIDR:-10.96.0.0/12}"
+    DIND_SUBNET_SIZE="${DIND_SUBNET_SIZE:-16}"
+    dns_server="${REMOTE_DNS64_V4SERVER:-8.8.8.8}"
+fi
+dns_prefix=$(echo $SERVICE_CIDR | sed 's,/.*,,')
+if [[ $IP_MODE = "ipv6" ]]; then
```
Comment (pmichali) [inline dind-cluster.sh]: Typo here. Should be !=, so that it pulls the zero off prefix (10.96.0.0/12 -> 10.96.0.0 -> **10.96.0.**)

File: dind-cluster.sh [hunk 7]
```diff
@@ -459,7 +561,7 @@ function dind::kubeadm {
   status=0
   # See image/bare/wrapkubeadm.
   # Capturing output is necessary to grab flags for 'kubeadm join'
-  if ! docker exec "${container_id}" wrapkubeadm "$@" 2>&1 | tee /dev/fd/2; then
+  if ! docker exec "${container_id}" wrapkubeadm ${dns_ip} "$@" 2>&1 | tee /dev/fd/2; then
```
Comment (pmichali) [inline dind-cluster.sh]: Unfortunately, this breaks CI, as running the new dind-cluster.sh, but using an older kubeadm-cluster-dind image, the wrapkubeadm on that image cannot deal with another argument. I'm looking into other options, like setting up the needed KUBELET_DNS_ARGS entry for systemd from the dindnet script, instead of passing into wrapkubeadm.

File: image/wrapkubeadm [hunk 8]
```diff
@@ -327,6 +337,9 @@ function dind::do-kubeadm {
   fi
 }
 
+dns_ip=$1
```
Comment (pmichali) [inline image/wrapkubeadm]: As mentioned above, we can't rely on passing in a new argument, as old images will not work with new dind-cluster.sh. So, we can't really do the changes on L306-314 here, as we have no easy way to configure the DNS IP on the fly.

File: README.md (hunk 2 above)
Comment (ivan4th) [reply README.md]: Indeed, after we merge the PR these commands are redundant. It will be possible to use any up-to-date DIND image for IPv6.

File: image/snapshot [hunk 9]
```diff
@@ -42,8 +42,7 @@ function snapshot::prepare {
     while kubectl get pod kube-controller-manager-kube-master -n kube-system -o name 2>/dev/null | grep -q pods*/; do
       if ((--n == 0)); then
         echo "WARNING: kubelet glitch: kube-controller-manager won't stop; pods may 'blink' for some time after restore" >&2
-        systemctl stop kubelet docker
```
Comment (ivan4th) [inline image/snapshot]: It's important to stop kubelet and docker before making the snapshot, otherwise it can turn out to be inconsistent.

File: dind-cluster.sh [hunk 10]
```diff
@@ -856,8 +985,12 @@ function dind::do-run-e2e {
   local parallel="${1:-}"
   local focus="${2:-}"
   local skip="${3:-}"
+  local host=$kube_master_ip
+  if [[ $IP_MODE = "ipv6" ]]; then
```
Comment (ivan4th) [inline dind-cluster.sh]: `[[ ${IP_MODE} = "ipv6" ]]`

File: dind-cluster.sh [hunk 11]
```diff
@@ -856,8 +985,12 @@ function dind::do-run-e2e {
   local parallel="${1:-}"
   local focus="${2:-}"
   local skip="${3:-}"
+  local host=$kube_master_ip
```
Comment (ivan4th) [inline dind-cluster.sh]: host="${kube_master_ip}"

File: dind-cluster.sh [hunk 12]
```diff
@@ -607,7 +734,7 @@ function dind::join {
 }
 
 function dind::escape-e2e-name {
-    sed 's/[]\$*.^|()[]/\\&/g; s/\s\+/\\s+/g' <<< "$1" | tr -d '\n'
```
Comment (ivan4th) [inline dind-cluster.sh]: Why this change?

File: dind-cluster.sh [hunk 13]
```diff
@@ -557,12 +669,27 @@ function dind::init {
     # insecure-bind-address and insecure-bind-port are also overridden by patching
     # apiserver static pod, but someday when k-d-c will only support
     # kubeadm 1.8+ we'll be able to remove that patching
+    local pod_net_cidr=""
+    # TODO: May want to specify each of the plugins that require --pod-network-cidr
+    if [[ ${CNI_PLUGIN} != "bridge" ]]; then
+      pod_net_cidr="podSubnet: \"${POD_NETWORK_CIDR}\""$'\n'"  "
+    fi
+    local bind_address="0.0.0.0"
+    if [[ $IP_MODE = "ipv6" ]]; then
```
Comment (ivan4th) [inline dind-cluster.sh]: `[[ ${IP_MODE} = "ipv6" ]]`

File: README.md [hunk 14]
```diff
@@ -123,6 +123,63 @@ The first `dind/dind-cluster.sh up` invocation can be slow because it
...
+be cherry picked and added to the repo:
+
+```
+PR #47621 "Updates RangeSize error message and tests for IPv6"
+PR #48551 "Add IPv6 support to iptables proxier"
+PR #50929 "Add kubeadm config for setting kube-proxy bind address"
+PR #50478 "Fix kube-proxy to use proper iptables commands for IPv6"
+PR #52028 "Add required family flag for conntrack IPv6 operation"
+PR #52033 "Removed the IPv6 prefix size limit for cluster-cidr"
+```
```
Comment (ivan4th) [inline README.md]: It would be nice to add some handy commands that may help with cherry-picking the PR, such as `git fetch origin pull/PR_ID/head:LOCAL_BRANCH`

File: config.sh (hunk 3 above)
Comment (ivan4th) [reply config.sh]: `EMBEDDED_CONFIG` actually disallows loading of the config:
```
if [[ ! ${EMBEDDED_CONFIG:-} ]]; then
  source "${DIND_ROOT}/config.sh"
fi
```

The idea of `config.sh` is to set env vars for `dind-cluster.sh` (but not `dind-cluster-vXXX.sh`, i.e. ones that use `EMBEDDED_CONFIG`)

File: dind-cluster.sh [hunk 15]
```diff
@@ -390,6 +432,55 @@ function dind::ensure-volume {
...
+		   resystit/bind9:latest >/dev/null
+	    ipv4_addr=$(docker exec bind9 ip addr list eth0 | grep "inet" | awk '$1 == "inet" {print $2}')
+	    docker exec bind9 ip addr del $ipv4_addr dev eth0
+	    docker exec bind9 ip -6 route add $DNS64_PREFIX_CIDR via $LOCAL_NAT64_SERVER
+        fi
+    fi
+}
+
+function dind::ensure-nat {
+    if [[  $IP_MODE = "ipv6" ]]; then
```
Comment (ivan4th) [inline dind-cluster.sh]: please use `[[ ${IP_MODE} = "ipv6" ]]` for the sake of consistency

File: dind-cluster.sh [hunk 16]
```diff
@@ -390,6 +432,55 @@ function dind::ensure-volume {
...
+        if ! docker ps | grep tayga >&/dev/null; then
+            docker run -d --name tayga --hostname tayga --net kubeadm-dind-net --label mirantis.kubeadm_dind_cluster \
+		   --sysctl net.ipv6.conf.all.disable_ipv6=0 --sysctl net.ipv6.conf.all.forwarding=1 \
+		   --privileged=true --ip 172.18.0.200 --ip6 $LOCAL_NAT64_SERVER --dns $REMOTE_DNS64_V4SERVER --dns $dns_server \
+		   -e TAYGA_CONF_PREFIX=$DNS64_PREFIX_CIDR -e TAYGA_CONF_IPV4_ADDR=172.18.0.200 \
+		   danehans/tayga:latest >/dev/null
+	    # TODO Way to add route w/o sudo? Need to check/create, as "clean" may remove route
+	    local route=$(ip route | egrep "^172.18.0.128/25")
```
Comment (ivan4th) [inline dind-cluster.sh]: `local route="$(ip route | egrep "^172.18.0.128/25")"`

File: dind-cluster.sh [hunk 17]
```diff
@@ -390,6 +432,55 @@ function dind::ensure-volume {
...
+	    if [[ -z "$route" ]]; then
```
Comment (ivan4th) [inline dind-cluster.sh]: `[[ ! ${route} ]]`

File: dind-cluster.sh [hunk 18]
```diff
@@ -479,7 +585,11 @@ function dind::kubeadm {
 
 function dind::configure-kubectl {
   dind::step "Setting cluster config"
-  "${kubectl}" config set-cluster dind --server="http://localhost:${APISERVER_PORT}" --insecure-skip-tls-verify=true
+  local host="localhost"
+  if [[ $IP_MODE = "ipv6" ]]; then
```
Comment (ivan4th) [inline dind-cluster.sh]: `[[ ${IP_MODE} = "ipv6" ]]`

**Comment by ivan4th** [comment]
As of now, the builds fail with this:
```
cp: cannot stat '/config-files/20-extra-dns-args.conf': No such file or directory
```

File: README.md (hunk 14 above)
Comment (pmichali) [reply README.md]: Will add some. Wasn't sure if you wanted the README with that much detail.

File: config.sh (hunk 3 above)
Comment (pmichali) [reply config.sh]: I moved DIND_SUBNET up to dind-cluster.sh, in order to avoid setting EMBEDDED_CONFIG for IPv6 mode (otherwise, it gets set by config.sh to IPv4 address. Will that work for the older scripts too? It only sets, if the value is not set already.

File: dind-cluster.sh (hunk 15 above)
Comment (pmichali) [reply dind-cluster.sh]: OK

File: dind-cluster.sh (hunk 16 above)
Comment (pmichali) [reply dind-cluster.sh]: Will update.

File: dind-cluster.sh (hunk 17 above)
Comment (pmichali) [reply dind-cluster.sh]: Will do for all.

File: dind-cluster.sh (hunk 18 above)
Comment (pmichali) [reply dind-cluster.sh]: ditto

File: dind-cluster.sh (hunk 13 above)
Comment (pmichali) [reply dind-cluster.sh]: ditto

File: dind-cluster.sh (hunk 12 above)
Comment (pmichali) [reply dind-cluster.sh]: We wanted to run a test like this, and with the escaping, it was not forming the right focus/skip in the underlying test run AFAICT:

```
../dind-cluster.sh e2e "Networking|Services" "IPv4|Networking-Performance|Federation|preserve source pod"
```

File: dind-cluster.sh (hunk 11 above)
Comment (pmichali) [reply dind-cluster.sh]: will do

File: dind-cluster.sh (hunk 10 above)
Comment (pmichali) [reply dind-cluster.sh]: ditto

File: image/snapshot (hunk 9 above)
Comment (pmichali) [reply image/snapshot]: Right. Instead of returning, we are doing a break, which falls through to L61 and will do the stop. Should be equivalent operation, right?

File: image/wrapkubeadm [hunk 19]
```diff
@@ -49,7 +49,7 @@ apiserver_streaming_proxy_redirects='.spec.containers[0].command|=.+["--feature-
 # Update kube-proxy CIDR, enable --masquerade-all and disable conntrack (see dind::frob-proxy below)
 function dind::proxy-cidr-and-no-conntrack {
   cluster_cidr="$(ip addr show docker0 | grep -w inet | awk '{ print $2; }')"
-  echo ".items[0].spec.template.spec.containers[0].command |= .+ [\"--cluster-cidr=${cluster_cidr}\", \"--masquerade-all\", \"--conntrack-max=0\", \"--conntrack-max-per-core=0\"]"
+  echo ".items[0].spec.template.spec.containers[0].command |= .+ [\"--cluster-cidr=${cluster_cidr}\", \"--masquerade-all\", \"--conntrack-max=-1\", \"--conntrack-max-per-core=-1\"]"
```
Comment (pmichali) [inline image/wrapkubeadm]: Issue 50787 alludes to being able to set these to -1 (rather than hack kube-proxy code). Just tried it, but it is not working. Investigating.

Comment (pmichali) [reply image/wrapkubeadm]: Looks like command line args are deprecated. In addition, they are ignored, if there is a config file (there is). Using a Kuberentes code hack to avoid this issue by diabling defaulting logic from changing conntrack values, if they are not specified in config file (they are not), which on some systems causes a problem as hashsize cannot be changed.

In future, will need to configure config file for kube-proxy with the needed values to prevent override (or get issue 25543 fixed).

**Comment by pmichali** [comment]
@ivan4th I updated the change set, based on your comments. Can you recheck?

Can you elaborate on the "build failure"? That file is not used in the building of the image (which is needed to get all the changes). It is created by dindnet, and then applied (copied into systemd area and then service restarted) during wrapkubeadm operation.

Thanks.

File: README.md (hunk 2 above)
Comment (pmichali) [reply README.md]: Do we want to leave this in for now, and then remove, once the PR is committed, or pull it now?

File: config.sh (hunk 3 above)
Comment (ivan4th) [reply config.sh]: Maybe a better way is to use separate `DIND_SUBNET_6` and `DIND_SUBNET_4` vars? The idea of `config.sh` is to provide the list of available knobs that customize k-d-c behaviour, and removing a var from there makes its usage unclear.

File: dind-cluster.sh (hunk 12 above)
Comment (ivan4th) [reply dind-cluster.sh]: Ok.

File: image/snapshot (hunk 9 above)
Comment (ivan4th) [reply image/snapshot]: Yes, sorry.

File: test.sh [hunk 20]
```diff
@@ -257,6 +257,10 @@ function test-case-src-master-weave {
   )
 }
 
+function test-case-src-working-area {
```
Comment (ivan4th) [inline test.sh]: What is this used for?

File: config.sh (hunk 3 above)
Comment (pmichali) [reply config.sh]: It seems like config.sh does more of forcing values to defaults, when not customizing, than showing all the available knobs (several are missing). This is important, as it forces one to customize settings to do IPv6 mode.

With the change set, one can do IPv6 mode, by setting the IP_MODE settings, and will get an IPv6 DIND_SUBNET default value, which can still be overridden, if needed.

If we use multiple variable names for the DinD subnet(we'd need 4, net and size for each IP mode), then we have to add conditional logic, based on the IP mode, to the network create, as well. I think splitting the variable up makes it more confusing.

Should we add IP_MODE to config.sh, defaulting to ipv4, and have logic there that sets DIND_SUBNET default, based on IP mode? IOW, move some of the dind_cluster.sh logic into config.sh?

File: test.sh (hunk 20 above)
Comment (pmichali) [reply test.sh]: Intent is to give the developer a way to run the CI test locally, against their customized local DinD and kubernetes repos. For example, for IPv6, there are numerous PRs in-flight, so we have a Kubernetes repo based on master, with a dozen PRs added.  Intent is not to add this as a test-case for a CI run.

**Comment by pmichali** [comment]
I’ll be pushing up a new commit that has several changes. One is to try to address DIND_SUBNET so that this can be used with IPv6 w/o customizing with EMBEDDED_CONFIG. Let me know what you think.

I have test-case-src-working-area as a way for developers to test the CI operation locally, with a custom Kubernetes repo. For example, I have a repo based on master, but with almost a dozen IPv6 in-flight PRs patched in. Intent is not to use this test case for CI, but for development use. Let me know if you want it removed or commented out.

My PR is based on the fix-1.8+ branch. Do I need to do anything, like rebase to a different branch? I wanted the fix-1.8+ changes. Also, my latest code (that I’m looking to update in the PR), has the workaround change you did for Travis. Is that committed anywhere? fix-1.8+?

I’ve been testing my changes and IPv4/IPv6 seem to work just fine locally. For GCE, I made a gcd-setup-local.sh script that does a build, so that a local DinD image can be used. It works for IPv4, but is not working for IPv6. I had to do several changes, so that files are over on the VM. Seems like it is trying to do kubectl command to start up the dashboard pod, and that request is not getting to the kube-master. It was trying to access localhost:8899, but localhost is 127.0.0.1.

With the latest changes, I tried doing SSH forward of ::1 using port 8899, but it is still failing. Not sure what is wrong there.  Would appreciate any thoughts you have on how to get this going.

Regards,

Paul Michali (pcm)

File: dind-cluster.sh [hunk 21]
```diff
@@ -483,7 +597,15 @@ function dind::kubeadm {
 
 function dind::configure-kubectl {
   dind::step "Setting cluster config"
-  "${kubectl}" config set-cluster dind --server="http://localhost:${APISERVER_PORT}" --insecure-skip-tls-verify=true
+  local host="[${dind_ip_base}2]"
+  if [[ "${GCE_HOSTED}" = true ]]; then
+    if [[ "${IP_MODE}" = "ipv4" ]]; then
+      host="localhost"
+    else
+      host="[::1]"
```
Comment (pmichali) [inline dind-cluster.sh]: For IPv6 GCE setup, kubectl commands are not making it to kube-master.

**Comment by pmichali** [comment]
Have both IPv4 and IPv6 working for bare-metal and GCE configurations.  Removed WIP from title. Please review.

File: dind-cluster.sh (hunk 21 above)
Comment (pmichali) [reply dind-cluster.sh]: works now.

**Comment by pmichali** [comment]
@ivan4th @danehans Please review so we can upstream this change. Thanks!

**Comment by pmichali** [comment]
Thanks for reviewing and approving @ivan4th !

**Comment by danehans** [comment]
Paul,

Do I call gce-setup-local.sh or gce-setup.sh to deploy to gce with ipv6?

-Daneyon

**Comment by pmichali** [comment]
I only tested with gce-setup-local.sh, because the image for IPv6 was not released. To do that, I needed to set DIND_CLUSTER and BUILD_* flags.  The key here is that the GCE_HOSTED flag needs to be set. If the IPv6 version is now released, we can revise gcd-setup.sh to set the GCE_HOSTED flag and then it would work for IPv6 too. Note: for testing, you could just set the environment variable, and then run gcd-setup.sh and it should work (provided the released image includes the IPv6 stuff).

So, for IPv6 we need:

export BUILD_KUBEADM=y
export BUILD_HYPERKUBE=y
export DIND_IMAGE=mirantis/kubeadm-dind-cluster:local
export IP_MODE=ipv6
. ../gce-setup-local.sh

Or (until updated)…

export GCE_HOSTED=true
export IP_MODE=ipv6
. ../gce-setup.sh

Assuming your Kubernetes area is directly below the DinD directory.

Regards,

Paul Michali (pcm)
